costmap_2d/plugins/obstacle_layer.cpp

```diff
@@ -444,11 +452,58 @@ void ObstacleLayer::updateBounds(double robot_x, double robot_y, double robot_ya
   updateFootprint(robot_x, robot_y, robot_yaw, min_x, min_y, max_x, max_y);
 }
 
+void ObstacleLayer::computeFootprintCenter(const std::vector<geometry_msgs::Point>& footprint)
+{
+    double sum_x = 0.0, sum_y = 0.0;
+    for (const auto& point : footprint)
+    {
+        sum_x += point.x;
+        sum_y += point.y;
+    }
+
+    footprint_center_.x = sum_x / footprint.size();
+    footprint_center_.y = sum_y / footprint.size();
+}
+
+std::vector<geometry_msgs::Point> ObstacleLayer::getReducedFootprint(const std::vector<geometry_msgs::Point>& original_footprint, double reduction_size)
+{
+    std::vector<geometry_msgs::Point> reduced_footprint;
+
+    for (const auto& p : original_footprint)
+    {
+        double dx = p.x - footprint_center_.x;
+        double dy = p.y - footprint_center_.y;
+
+        double distance = sqrt(dx * dx + dy * dy);
+        geometry_msgs::Point new_point;
+
+        if (distance > 0)
+        {
+            double scale = (distance - reduction_size) / distance;
```

Comment:
scale = 1 - red_size/distance right? Is check for valid values needed ?

Reply:
if distance is 0 (however it is only possible if centroid of footprint is same as corner point of footprint) it will crash because it will divide by 0. So I have added the check for safety.

Reply:
Q: can't we change the signature of the method to be the scale? I think it is more intuitive to has `scale` directly as argument and it should be flexible to increase / decrease (can be > 1).
btw, can't we use `costmap2d::padFootprint`?